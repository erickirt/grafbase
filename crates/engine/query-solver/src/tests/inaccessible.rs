use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/union-intersection/a")
    B @join__graph(name: "b", url: "http://localhost:4200/union-intersection/b")
}

type Book
    @join__type(graph: A)
    @join__type(graph: B)
{
    title: String!
}

type Movie
    @join__type(graph: B)
{
    title: String!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
{
    media: Media
    book: Media @join__field(graph: A, type: "Book") @join__field(graph: B, type: "Media")
    song: Media @join__field(graph: A)
    viewer: Viewer
}

type Song
    @join__type(graph: A)
{
    title: String!
}

type Viewer
    @join__type(graph: A)
    @join__type(graph: B)
{
    media: ViewerMedia
    book: ViewerMedia @join__field(graph: A, type: "Book") @join__field(graph: B, type: "ViewerMedia")
    song: ViewerMedia @join__field(graph: A)
}

union Media
    @join__type(graph: A)
    @join__type(graph: B)
    @join__unionMember(graph: A, member: "Book")
    @join__unionMember(graph: B, member: "Book")
    @join__unionMember(graph: A, member: "Song")
    @join__unionMember(graph: B, member: "Movie")
 = Book | Song | Movie

union ViewerMedia
    @join__type(graph: A)
    @join__type(graph: B)
    @join__unionMember(graph: A, member: "Book")
    @join__unionMember(graph: B, member: "Book")
    @join__unionMember(graph: A, member: "Song")
    @join__unionMember(graph: B, member: "Movie")
 = Book | Song | Movie
"#;

#[test]
fn inaccessible_field_for_one_subgraph() {
    // title is available in subgraph B not in A. Query.media is accessible in both.
    assert_solving_snapshots!(
        "inaccessible_field_for_one_subgraph",
        SCHEMA,
        r#"
        query {
          media {
            ... on Movie {
              title
            }
          }
        }
        "#
    );
}
