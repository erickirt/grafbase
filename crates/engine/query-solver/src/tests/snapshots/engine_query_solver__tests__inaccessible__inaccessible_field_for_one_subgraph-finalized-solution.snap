---
source: crates/engine/query-solver/src/tests/inaccessible.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#b\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"Query.media\" ]\n    3 [ label = \"Movie.title\" ]\n    0 -> 1 [ color=royalblue,fontcolor=royalblue ]\n    2 -> 3 [  ]\n    1 -> 2 [  ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#b" ]
    2 [ label = "Query.media" ]
    3 [ label = "Movie.title" ]
    0 -> 1 [ label = "QueryPartition" ]
    2 -> 3 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
}
